design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,163 @@
+
+```
+	// CPU allow specified the detailed CPU topology inside the vmi.
+		CurrentCPUTopology *CPU `json:"currentCPUTopology,omitempty"`
+```
+
+Conditions on the VM and VMI should indicate the status of the pending CPU hotplug actions. When errors occur, the conditions should propagate the errors to the VM.Status.Conditions field so users can receive feedback into the status of their declared state.
+
+```
+	// Indicates that the VMI is in progress of Hot vCPU Plug/UnPlug
+	VirtualMachineInstanceVCPUChange VirtualMachineInstanceConditionType = "HotVCPUChange"
+```
+
+Upon succesful hotplug action the condition will be removed. Otherwise, a human readable reason and message indicating whether the hotplug change is pending or encountering errors.
+
+## Extensibility
+We will be able to use the new `liveUpdateFeatures` section to manage things like memory hotplug in the future.
+
+## Update/Rollback Compatibility
+This is an entirely new feature, so updates from older versions will not be a problem.
+
+## Functional Testing Approach
+Functional tests will simply:
+- Create and start a VM with `cpu` set as one of the `liveUpdateFeatures`
+- Modify the number of CPU sockets for the VM
+- Ensure the guest sees the changed number of CPUs
+- Ensure the CPU resources on the virt-launcher pod were adjusted accordingly
+
+# Implementation Phases
+
+The first two phases will happen together in the same pull request (PR), since the first phase is mostly useless without the second one.
+In the unlikely case that hot-unplugging CPUs is more complicated than anticipated, hotplug-only could be submitted as a first PR, then hot-unplug later on as a second PR.
+The third phase will happen later as a separate PR. It should be able to transparently replace the migration mechanism without API changes.
+
+## Phase 1: Creating the KubeVirt API and linking it to the LibVirt CPU hotplug API
+## Phase 2: Triggering a live migration when the number of sockets changes, to adjust the virt-launcher CPU resources
+
+The VMI [workload update controller](https://github.com/kubevirt/kubevirt/blob/ce8aab7874d2e1586787e3e2a17306b7edca1b8a/pkg/virt-controller/watch/workload-updater/workload-updater.go#L476) is extended to live migrate VMIs to satisfy updates to a VMI resource request/limits field.
+
+Sequence of events:
+- User updates CPU sockets on a vm.spec.template.spec.domain.cpu
+- VM controller determines the request/limits required and writes the necessary changes to the active vmi.Spec
+- VMI workload update controller detects the mismatch between the resource request/limits on the pod vs vmi.Spec and live migrates the VMI to satisfy vmi.spec
+- Migration controller constructs a new target pod with the desired CPU request/limits
+- VMI live migrates
+- virt-handler triggers the hotplug/hot-unplug LibVirt action after the live migration completes
```

Comment:
I completely  miss an information which will be used by handler to determine that the migration happened.

Reply:
Hmm.. this should take place in finalizeMigration() - this assumes that the migration is over.
Here virt-handler would check if vmi.Status.CurrentCPUTopology exists and will call `SyncVirtualMachineCPUs`

Reply:
but this seems like an implementation detail to me.

Reply:
I agree that it is implementation detail but at the same time one of which matters imho. Why? Well as you said in this case you would like to implement it in `finalizeMigration` but what about the version where we use in-place update? 

Also `SyncVirtualMachineCPUs` seems like new API and it is not documented. Why do we need it and why can't we do this in `SyncVMI` (existing) call?

Reply:
> what about the version where we use in-place update?

such a version doesn't exist today. However, `SyncVirtualMachineCPUs` was introduced specifically to address the vCPU change which can happen with or without migration.
I honestly don't see how we can mix this functionality with a regular syncVMI flow. 
`SyncVirtualMachineCPUs` should happen at a specific time when we know that vCPUs can be manipulated. virt-handler should orchestrate this step and it shouldn't happen on ever syncVMI call.

Reply:
>such a version doesn't exist today. However, SyncVirtualMachineCPUs was introduced specifically to address the vCPU change which can happen with or without migration.

From goals section "Underlying technology used for hotplug should be independent of the VM API, meaning we should be able to swap out how hotplug is technically performed in the future without impacting API compatibility"

I am trying to see if current approach/API is sufficient for seamless change of way we do this.

Therefore let me re-iterate once again. How do we signal the hotplug in case of different way of implementing this?

While writing this I also don't understand what is an expecting behavior in this case:
1. VM is being migrated or is going to be migrated
2. Users changes the number of sockets
3. Migration is completed

What happens in step 3? Will we explode?

Reply:
@enp0s3 Consistency and it should be rather other way around. What is benefit of new API and not keeping it consistent?

Reply:
@acardace 

>I agree with @vladikr on this, we need a new API to make this easier to work with when the vertical scaling API will land.

I am probably missing something. How is this making it easier? Also please document the API requirement here.

>SyncVMI is not a good place to put hotplug logic otherwise you will see continuous notification of a device being hotplugged in the guest, it's definitely better to hotplug when requested with a different code path.

Why would we have continuous notifications? Would simple `if` be sufficient?
`if hasDifferentCPU {
    do the magic and log, otherwise no notification
}`

>it's definitely better to hotplug when requested with a different code path.

What are the benefits?

Reply:
> I am probably missing something. How is this making it easier? Also please document the API requirement here.

I mean internal API, between libvirt and kubevirt.

Requirement is it should hotplug CPUs :)

> 
> > SyncVMI is not a good place to put hotplug logic otherwise you will see continuous notification of a device being hotplugged in the guest, it's definitely better to hotplug when requested with a different code path.
> 
> Why would we have continuous notifications? Would simple `if` be sufficient? `if hasDifferentCPU { do the magic and log, otherwise no notification }`

Yes, of course, you can do it by adding fields in the vmi.Status or check what the current domain looks like, to check whether you already hotplugged CPUs or whether a hotplug is in progress but SyncVMI() creates a new api.Domain anew every time and I'd rather leave its functionality and semantics as is.

I think it's simpler and more maintainable/readable if we use another internal API to do live hotplug which is a one-time live operation.

> 
> > it's definitely better to hotplug when requested with a different code path.
> 
> What are the benefits?

For me simplicity.

Reply:
> > such a version doesn't exist today. However, SyncVirtualMachineCPUs was introduced specifically to address the vCPU change which can happen with or without migration.
> 
> From goals section "Underlying technology used for hotplug should be independent of the VM API, meaning we should be able to swap out how hotplug is technically performed in the future without impacting API compatibility"
> 
> I am trying to see if current approach/API is sufficient for seamless change of way we do this.
> 
> Therefore let me re-iterate once again. How do we signal the hotplug in case of different way of implementing this?
> 
> While writing this I also don't understand what is an expecting behavior in this case:
> 
> 1. VM is being migrated or is going to be migrated
> 2. Users changes the number of sockets
> 3. Migration is completed
> 
> What happens in step 3? Will we explode?

After the VM has been migrated virt-controller will notice that the CPUTopology in the vmi.Status is different than the desired one and will hotplug I guess?

Reply:
> > I mean internal API, between libvirt and kubevirt.
> 
> I don't understand the answer. We never strived to have 1:1 mapping to libvirt nor do I think it is easier.

I didn't say that.

> 
> > Yes, of course, you can do it by adding fields in the vmi.Status or check what the current domain looks like, to check whether you already hotplugged CPUs or whether a hotplug is in progress but SyncVMI() creates a new api.Domain anew every time and I'd rather leave its functionality and semantics as is.
> 
> I believe the semantics doesn't change.

I don't agree.

> 
> > I think it's simpler and more maintainable/readable if we use another internal API to do live hotplug which is a one-time live operation.
> 
> My experience is contradicting this. The less we expand this internal API the better as we reuse what we already have. Also now we are leaking logic which is responsible for what should happen to virt-handler. My perception is that we mostly kept this in virt-launcher.

I don't know maybe you're right about all of this, it's also hard to discuss this back and forth in GH like this, what's most important is that I think we're diverging from the objective, this is a design PR, not an implementation one.

What I mean is that I don't really mind where it lands (SyncVMI vs separate API) as long as the implementation is sound, but we should focus on the overall design here which I think it's outlined pretty clearly.

> 
> > After the VM has been migrated virt-controller will notice that the CPUTopology in the vmi.Status is different than the desired one and will hotplug I guess?
> 
> Sure but migrated Pod will have the same resources, meaning not adjusted to what was asked, and still I believe you want to trigger the hot un/plug. Leaving the domain updated but Pod is not. Or even worse making race behavior.

Then it's going to migrate again, but I'd stop here since I feel like we're going into a rabbit hole and these are implementation details anyway, the desired outcome is clear, hotplug has to happen if requested.

Reply:
> > I mean internal API, between libvirt and kubevirt.
> 
> I don't understand the answer. We never strived to have 1:1 mapping to libvirt nor do I think it is easier.
> 
> > Yes, of course, you can do it by adding fields in the vmi.Status or check what the current domain looks like, to check whether you already hotplugged CPUs or whether a hotplug is in progress but SyncVMI() creates a new api.Domain anew every time and I'd rather leave its functionality and semantics as is.
> 
> I believe the semantics doesn't change.
> 
> > I think it's simpler and more maintainable/readable if we use another internal API to do live hotplug which is a one-time live operation.
> 
> My experience is contradicting this. The less we expand this internal API the better as we reuse what we already have. Also now we are leaking logic which is responsible for what should happen to virt-handler. My perception is that we mostly kept this in virt-launcher.
> 
> > After the VM has been migrated virt-controller will notice that the CPUTopology in the vmi.Status is different than the desired one and will hotplug I guess?
> 
> Sure but migrated Pod will have the same resources, meaning not adjusted to what was asked, and still I believe you want to trigger the hot un/plug. Leaving the domain updated but Pod is not. Or even worse making race behavior.

@xpivarc It will not. The target pod will have adjusted resources. 
When we create the target pod, it is being created from a template that uses the updated the vmi spec, thus it will correctly calculate and update the requested resources.